### PR TITLE
Stop allowing accented characters in tags

### LINKED ISSUE #8884: Tag limitation
**Describe the bug**

I was able to create a tag with only one special character `â` when adding a new post. But the generated page with tag did not work.

**To Reproduce**

1. Write a new post
2. Add a tag with some special character like `â` or `ą` or whatever else.. Technically you can't enter that character by hand (usually with use of `opt` key on keyboard) but you can easily paste it or long press `a` key to see diacritical marks and then you can enter any you one.
3. Post is created. Click on that tag in the post view.
4. Error.

**Video**
[Video](https://d.pr/v/1xSmcP)

**Expected behavior**

Not sure... 🤷 

**Desktop (please complete the following information):**

- OS: macOS X
- Browser: Chrome
- Version: 83.x

### PR DESCRIPTION
**Pull request by jacobherrington**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This commit will add strict frontend and backend validation for tags
disallowing accented characters.

## Related Tickets & Documents

Closes #8884

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

Before:
https://d.pr/v/1xSmcP

After:
Peek 2020-06-29 20-54

*Note: HTML5 validation won't let you type accented characters, but you can paste them*

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: spec/models/tag_spec.rb [hunk 1]
```diff
@@ -46,6 +46,16 @@
...
         expect(tag).not_to be_valid
       end
+
+      it "fails validations if name uses diacritics" do
+        tag.name = "łookatmé"
+        expect(tag).not_to be_valid
+      end
+
+      it "fails validations if name uses non-latin characters" do
+        tag.name = "火"
```
Comment (jacobherrington) [inline spec/models/tag_spec.rb:56]: We could certainly expand this to include more character sets, but I think that might be overdoing it. :stuck_out_tongue:

Comment (citizen428) [reply spec/models/tag_spec.rb:56]: I don't think so tbh. I left a comment on the issue, but given our coming I18n efforts it seems restricting tags to ASCII is a bit counterproductive.

File: app/models/tag.rb [hunk 2]
```diff
@@ -82,7 +82,10 @@ def self.find_preferred_alias_for(word)
 
   def validate_name
     errors.add(:name, "is too long (maximum is 30 characters)") if name.length > 30
-    errors.add(:name, "contains non-alphanumeric characters") unless name.match?(/\A[[:alnum:]]+\z/)
+    # [:alnum] is not used here because it supports diacritical characters.
+    # If we decide to allow diacritics in the future, we should replace the
+    # following regex with [:alnum].
+    errors.add(:name, "contains non-alphanumeric characters") unless name.match?(/\A[[a-z0-9]]+\z/i)
```
Comment (citizen428) [inline app/models/tag.rb:88]: We should consider changing the wording to "contains non-ASCII characters". Accented characters are alphanumeric too, which is why the `:alnum:` class matches them.

**Comment by atsmith813** [comment]
WRT [`acts-as-taggable-on` gem](https://github.com/mbleigh/acts-as-taggable-on), this looks solid, thanks for jumping on this 🎉 . Since tags have been cleaned up, this will be much easier to manage now and in the future 💯 .

The gem offers [minor support for special characters for MySql users](https://github.com/mbleigh/acts-as-taggable-on#for-mysql-users) (which doesn't apply since we use PostgreSQL).

File: spec/models/tag_spec.rb (hunk 1 above)
Comment (jacobherrington) [reply spec/models/tag_spec.rb:56]: @citizen428 I agree in the long run.

Comment (jacobherrington) [reply spec/models/tag_spec.rb]: Lol re-reading this comment, I meant adding more tests was overdoing it... :sweat_smile: 

I think we should work toward removing this limitation on tags, especially considering that in the future there might be non-English communities on Forem.

**Comment by citizen428** [comment]
> The gem offers minor support for special characters for MySql users (which doesn't apply since we use PostgreSQL).

I've maintained for a long time that it's almost always better to implement tagging from scratch. It's not particularly complicated and over the years I found that existing solutions almost always fall short and get replaced anyway, same as admin scaffolding tools (see recent discussion about removing Administrate).

**Comment by djuber** [comment]
https://github.com/forem/forem/issues/14422 adding this issue link for visibility into the context (and #13720 as well)

**Pull request merged**